solidity/contracts/BeaconBackportRewards.sol

```diff
@@ -0,0 +1,141 @@
+        // that should be excluded from the rewards.
+        uint256[] memory _excludedGroups,
+        // The address that receives any left-over tokens
+        // from late funding or excluded groups.
+        address _excessRecipient
+    ) public Rewards(
+        _termLength,
+        _token,
+        _minimumKeepsPerInterval,
+        _firstIntervalStart,
+        _intervalWeights
+    ) {
+        operatorContract = KeepRandomBeaconOperator(_operatorContract);
+        tokenStaking = TokenStaking(_stakingContract);
+        lastGroupOfInterval = _lastGroupOfInterval;
+        for (uint256 i = 0; i < _excludedGroups.length; i++) {
+            excludedGroups[_excludedGroups[i]] = true;
+        }
+        excessRecipient = _excessRecipient;
+    }
+
+    function lastEligibleGroup() public view returns (uint256) {
+        return lastGroupOfInterval[lastInterval()];
+    }
+
+    function lastInterval() public view returns (uint256) {
+        return lastGroupOfInterval.length.sub(1);
+    }
+
+    function withdrawExcess() mustBeFinished(lastInterval()) public {
+        if (lastInterval() >= intervalAllocations.length) {
+            allocateRewards(lastInterval());
+        }
+        token.safeTransfer(excessRecipient, unallocatedRewards);
+        unallocatedRewards = 0;
+    }
+
+    function _assignedInterval(bytes32 groupIndexBytes) internal view returns (uint256) {
+        for (uint256 interval = 0; interval < lastGroupOfInterval.length; interval++) {
+            if (lastGroupOfInterval[interval] >= uint256(groupIndexBytes)) {
+                return interval;
+            }
+        }
+    }
+
+    function _getKeepCount() internal view returns (uint256) {
+        return lastEligibleGroup().add(1);
+    }
+
+    function _getKeepAtIndex(uint256 i) internal view returns (bytes32) {
+        return bytes32(i);
+    }
+
+    function _getCreationTime(bytes32 groupIndexBytes) internal view returns (uint256) {
+        // Assign each group to the starting timestamp of its interval
+        return startOf(_assignedInterval(groupIndexBytes));
+    }
+
+    function _isClosed(bytes32 groupIndexBytes) internal view returns (bool) {
+        if (_isTerminated(groupIndexBytes)) { return false; }
+        bytes memory groupPubkey = operatorContract.getGroupPublicKey(
+            uint256(groupIndexBytes)
+        );
+        return operatorContract.isStaleGroup(groupPubkey);
+    }
+
+    function _isTerminated(bytes32 groupIndexBytes) internal view returns (bool) {
+        return excludedGroups[uint256(groupIndexBytes)];
+    }
+
+    // A group is recognized if its index is at most `lastEligibleGroup`
+    // and it isn't listed as excluded.
+    function _recognizedByFactory(bytes32 groupIndexBytes) internal view returns (bool) {
+        return lastEligibleGroup() >= uint256(groupIndexBytes);
+    }
+
+    function _distributeReward(bytes32 groupIndexBytes, uint256 _value) internal {
```

Comment:
For each member, we do a token transfer which has a floor of 10k gas if the beneficiary has an existing balance, or 25k gas if the beneficiary has zero tokens. With 64 members this can be up to 1.6M gas. Reducing the beneficiaries to a map may not help much, as each new beneficiary still creates a 20k `SSTORE` operation. Optimization in the form of deduplicating member operators and then using the withdrawal pattern for token transfers, only querying the beneficiary address upon withdrawal, would slightly reduce the gas cost of calling `receiveReward` at the cost of possibly increasing the overall gas cost.

Reply:
If the reward for each keep is recorded only once, and verifiable members of the keep can withdraw at their leisure, this can be significantly optimized, actually.